File: Blind_75_List/DP - Multidimensional/1143. Longest Common Subsequence.cpp
//Problem Link: https://leetcode.com/problems/longest-common-subsequence/

class Solution {
    string s, t;
    int dp[1001][1001];
    int solve(int i, int j){
        if (i == s.size() || j == t.size())
                return 0;
        int& me = dp[i][j];
        if (~me)
            return me;
        if (s[i] == t[j]){
            return me = 1 + solve(i+1, j+1);
        }
        return me = max(
        solve(i+1, j),
        solve(i, j+1)
        );
    }
public:
    int longestCommonSubsequence(string text1, string text2) {
        s = text1;
        t = text2;
        memset(dp, -1, sizeof dp);
        return solve(0,0);
    }
};

File: Blind_75_List/DP - Multidimensional/1143. Longest Common Subsequence.py
# Problem Link: https://leetcode.com/problems/longest-common-subsequence/

class Solution:
    def longestCommonSubsequence(self, a: str, b: str) -> int:
        row = len(a)
        col = len(b)
        dp = [[0 for j in range(col+1)] for i in range(row+1)]
        for i in range(row-1,-1,-1):
            for j in range(col-1,-1,-1):
                if a[i] == b[j]:
                    dp[i][j] = 1 + dp[i+1][j+1]
                else:
                    dp[i][j] = max(dp[i+1][j], dp[i][j+1])
        
        
        
        
        return dp[0][0]